### PR TITLE
Import export data

### PR DESCRIPTION
**Pull request by leventdem**
Add import export

### CODE REVIEW

File: src/masq.js [hunk 1]
```diff
@@ -288,9 +284,124 @@ class Masq {
...
+
+    const derivedkey = await MasqCrypto.utils.deriveKey(passphrase, publicProfile.salt)
+    if (!derivedkey || derivedkey.length === 0) {
+      throw common.generateError(common.ERRORS.WRONGPASSPHRASE)
+    }
+    try {
+      const aesCipher = await new MasqCrypto.AES({key: derivedkey})
+      const masterKey = await aesCipher.unwrapKey(publicProfile.encryptedMasterKey, publicProfile.iv)
+
+      const dumpedStore = Object.assign({}, await this.getProfile())
```
Comment (G-Ray) [inline src/masq.js]: Only copy the needed field into a new object

File: src/masq.js [hunk 2]
```diff
@@ -288,9 +284,124 @@ class Masq {
...
+    this._checkCurrentUser()
+    if (!importRules || Object.keys(importRules).length === 0) {
+      throw common.generateError(common.ERRORS.NOIMPORTRULES)
+    }
+    const publicProfile = await this.getUser()
+    if (!passphrase || passphrase === '') {
+      throw common.generateError(common.ERRORS.NOPASSPHRASE)
+    }
+
+    const derivedkey = await MasqCrypto.utils.deriveKey(passphrase, publicProfile.salt)
```
Comment (G-Ray) [inline src/masq.js]: Like in a previous comment, can we throw the error from `deriveKey()` ?

File: src/masq.js [hunk 3]
```diff
@@ -288,9 +284,124 @@ class Masq {
...
+      throw common.generateError(common.ERRORS.WRONGPASSPHRASE)
+    }
+    try {
+      const aesCipher = await new MasqCrypto.AES({key: derivedkey})
+      const masterKey = await aesCipher.unwrapKey(publicProfile.encryptedMasterKey, publicProfile.iv)
+      const appList = await this.listApps()
+      const currentAppList = Object.keys(appList).map(key => appList[key].url)
+
+      for (let appUrl of Object.keys(importRules)) {
+        if (importRules[appUrl] === true) {
```
Comment (G-Ray) [inline src/masq.js]: ?

File: test/masq.js [hunk 4]
```diff
@@ -444,12 +444,84 @@ describe('Masq Store', () => {
...
+      expect(importRules).to.deep.equal({ 'https://www.qwant.com/music': true })
+      
+      
+      //Update exported data to simulate a new device
+      exportedStore[appUrl].foo = 'baz'
+      //import data in music application
+      await store.import(exportedStore, importRules, passphrase)
+      const res = await appData.getItem('foo')
+      expect(res).to.equal('baz')
+
```
Comment (G-Ray) [inline test/masq.js]: beware of empty lines :P

File: test/masq.js [hunk 5]
```diff
@@ -444,12 +444,84 @@ describe('Masq Store', () => {
...
+      expect(res).to.equal('baz')
+      
+
+      
+    })
+
+    it('Should delete the application ', async () => {
+      let appId = null
+      const apps = await store.listApps()
+      appId = Object.keys(apps)[0]
```
Comment (G-Ray) [inline test/masq.js]: use `const appId =` directly, and don't init the variable to null

File: src/masq.js [hunk 6]
```diff
@@ -288,9 +285,106 @@ class Masq {
...
+
+    const derivedkey = await MasqCrypto.utils.deriveKey(passphrase, publicProfile.salt)
+    try {
+      const aesCipher = await new MasqCrypto.AES({key: derivedkey})
+      const masterKey = await aesCipher.unwrapKey(publicProfile.encryptedMasterKey, publicProfile.iv)
+
+      const dumpedStore = {
+        appList: (await this.getProfile()).appList
+      }
+      for (let key of Object.keys(dumpedStore.appList)) {
```
Comment (G-Ray) [inline src/masq.js]: If you do not need the key, just iterate over `Object.values`

File: src/masq.js [hunk 7]
```diff
@@ -288,9 +285,106 @@ class Masq {
...
+      throw common.generateError(common.ERRORS.NOIMPORTRULES)
+    }
+    const publicProfile = await this.getUser()
+    this._checkPassphrase(passphrase)
+
+    const derivedkey = await MasqCrypto.utils.deriveKey(passphrase, publicProfile.salt)
+    try {
+      const aesCipher = await new MasqCrypto.AES({key: derivedkey})
+      const masterKey = await aesCipher.unwrapKey(publicProfile.encryptedMasterKey, publicProfile.iv)
+      for (let appUrl of Object.keys(importRules)) {
```
Comment (G-Ray) [inline src/masq.js]: Just iterate over `Object.values`

Comment (leventdem) [reply src/masq.js]: changed

**Pull request merged**